doc/developer/design/20220516_transactional_consistency.md

```diff
@@ -0,0 +1,449 @@
+transaction is committing, no other queries can execute. Once the commit is complete, all writes are guaranteed to be
+durable.
+
+#### Read Committed
+
+See [https://jepsen.io/consistency/models/read-committed](https://jepsen.io/consistency/models/read-committed) for an in
+depth formal discussion of Read Committed. Read Committed prevents Dirty Reads, which means that one transaction cannot
+view the uncommitted data of another transaction.
+
+This property holds from how Materialize satisfies [Read Uncommitted](#Read Uncommitted). No read can view uncommitted
+buffered data until after the commit is complete.
+
+#### Repeatable Read
+
+See [https://jepsen.io/consistency/models/repeatable-read](https://jepsen.io/consistency/models/repeatable-read) for an
+in depth formal discussion of Repeatable Reads. Repeatable Reads prevents Fuzzy/Non-Repeatable Reads, which means that
+once a transactions reads some value for an object, it will always read that same value, regardless if it’s been
+overwritten by another transaction.
+
+Differential Dataflow provides us with some useful properties to achieve this.
+
+1. All reads and writes take place at a provided timestamp.
+2. All reads at timestamp `ts` will return all writes at timestamp ≤ `ts`.
+3. All reads at timestamp `ts` will not return any of the writes at timestamps > `ts`.
+
+As described in the [Timestamp Selection](#Timestamp Selection) section, whenever a read happens at timestamp `ts`, all
+subsequent writes will happen at a timestamp > `ts`. Since a read transaction performs all of it’s read at the same
+timestamp, we can guarantee that subsequent reads will always read the same value since all writes will happen at a
+later timestamp.
+
+#### Serializable
+
+See [https://jepsen.io/consistency/models/serializable](https://jepsen.io/consistency/models/serializable) for an in
+depth formal discussion of Serializable. Serializability prevents Phantom Reads, which means that once a transaction
+reads some value that satisfy some predicate, it will always read the same values, regardless if some other transaction
+has made a write that affects the predicate. Informally this means that you can apply some total order to all
+transactions.
+
+This property holds from how Materialize satisfies [Repeatable Read](#Repeatable Read). All reads in a read only
+transaction happen at the same timestamp. All writes that happen after the first read of a read-only transaction will
+happen at a later timestamp, and therefore will not affect any reads of the read-only transaction.
+
+Informally we can use the transaction timestamps to provide a total ordering over all transactions.
+
+- Transaction T1 is ordered before Transaction T2 if the timestamp of T1 < the timestamp of T2.
+- Transaction T1 is ordered before Transaction T2 if the timestamp of T1 == the timestamp of T2, T1 is write-only, and
```

Comment:
I'm copying the `TimestampOracle` code here because it's a useful reference for this:
https://github.com/MaterializeInc/materialize/blob/43bad275b94a8e1f9fbd187f976bf18f7867ae3f/src/coord/src/coord.rs#L5441-L5476

This can happen with reads following a write, but not with writes following a read.

Consider a table t where:
- `upper` = `ts`
- `since` = `ts - 1`
- global timestamp = `ts`
- The most recent action was a write.

Then a client takes the following actions:
1. Writes some value to table t. The timestamp of this write will be `ts` and nothing is done to the global timestamp or table's `upper` and `since`.
2. Read from table t. The timestamp of this read is also `ts`. The global timestamp stays at `ts`, but `upper` is advanced to `ts + 1` (and `since` will likely advance to `ts` if there's no read holds).

Reply:
Oh derp--yep, that makes perfect sense.